### PR TITLE
Added analyzer for detecting recursive/nested mutex read locks.

### LINKED ISSUE #8323: [Feature Request] Add in a Custom Static Analyzer for Mutex Locks
# 🚀 Feature Request

### Description

Deadlocks while not common have come up more times than I would like. The majority of the times(if not all) this can be attributed to improper lock usage.

### Describe the solution you'd like

In the current form of how our current static analyzers are currently structured, we can have a similar one to prevent developers from inadvertently introducing recursive read locking/locking. Given that this is a deadlock condition that can eventually grind down our core beacon/validator processes to a halt, it is important to have a better way to detect errors such as these. 

By parsing the AST of the respective files in the repository, we can determine where mutex locks are being improperly applied and remedy them. 

For an example which illustrates this problem: 
#8314 
#8321 
#7905

Some examples of analyzers currently being applied in Prysm:
https://github.com/prysmaticlabs/prysm/tree/develop/tools/analyzers

### Describe alternatives you've considered

As far as I am aware there have been no external static linting tools/ analyzer to check for improper lock usage.

### PR DESCRIPTION
**Pull request by Heph789**
**What type of PR is this?**

> Uncomment one line below and remove others.
>
> Bug fix
Feature
> Documentation
> Other

**What does this PR do? Why is it needed?**
This analyzer detects recursive/nested mutex read locks.

**Which issues(s) does this PR fix?**

Fixes #8323

**Other notes for review**
I previously submitted this for review [here](https://github.com/prysmaticlabs/prysm/pull/8775), but I made a new branch because the commit history was starting to get messy. There are some known RLock patterns that the analyzer will not detect, as shown in recursivelock/testdata/globallocks.go and recursivelock/testdata/complexlocks.go. The bazel files may also need working on/fixing. Hopefully this tool is useful!

### CODE REVIEW

File: tools/analyzers/recursivelock/analyzer.go [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+				break
+			}
+			name := call.name
+			selMap := mapSelTypes(stmt, pass)
+			if selMap == nil {
+				break
+			}
+			if keepTrackOf.rLockSelector != nil {
+				if keepTrackOf.foundRLock > 0 {
+					if keepTrackOf.foundRLock > 0 && keepTrackOf.rLockSelector.isEqual(selMap, 0) {
```
Comment (rauljordan) [inline tools/analyzers/recursivelock/analyzer.go:75]: Can these two conditionals be simplified or merged?

File: tools/analyzers/recursivelock/analyzer.go [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+}
+
+// hasNestedRLock takes a call expression represented by callInfo as input and returns a stack trace of the nested or recursive RLock within
+// that call expression. If the call expression does not contain a nested or recursive RLock, hasNestedRLock returns an empty string.
+// hasNestedRLock finds a nested or recursive RLock by recursively calling itself on any functions called by the function/method represented
+// by callInfo.
+func hasNestedRLock(fullRLockSelector *selIdentList, compareMap *selIdentList, call *callInfo, inspect *inspector.Inspector, pass *analysis.Pass, hist map[string]bool) (retStack string) {
+	// debug := debugHelper{
+	// 	pass: pass,
+	// }
```
Comment (rauljordan) [inline tools/analyzers/recursivelock/analyzer.go:336]:
```suggestion
```

File: tools/analyzers/recursivelock/testdata/complexlocks.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
+// These nested rlock patterns are too complex for the analyzer to catch right now
```
Comment (nisdas) [inline tools/analyzers/recursivelock/testdata/complexlocks.go:1]: Given that there are certain classes of locks that this analyzer won't catch, do you mind opening an issue to track them ? This way if anyone wants to solve it and needs context they can fall back on the issue.

File: tools/analyzers/recursivelock/analyzer.go [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+		(*ast.FuncLit)(nil),
+		(*ast.File)(nil),
+		(*ast.ReturnStmt)(nil),
+	}
+
+	var keepTrackOf tracker
+	inspect.Preorder(nodeFilter, func(node ast.Node) {
+		if keepTrackOf.funcLitEnd.IsValid() && node.Pos() <= keepTrackOf.funcLitEnd {
+			return
+		} else {
```
Comment (nisdas) [inline tools/analyzers/recursivelock/analyzer.go:49]: No need for else here as  you are returning if the initial conditional is triggered.

File: tools/analyzers/recursivelock/analyzer.go [hunk 5]
```diff
@@ -0,0 +1,455 @@
...
+func (t *tracker) deincFRU() {
+	if t.foundRLock > 0 {
+		t.foundRLock -= 1
+	}
+}
+func (t *tracker) incFRU() {
+	t.foundRLock += 1
+}
+
+type selIdentNode struct {
```
Comment (nisdas) [inline tools/analyzers/recursivelock/analyzer.go:158]: Maybe some comments on why `selIdentNode` and `selIdentList` is needed and what they are used for. Also some comments on their respective methods and what they do would be helpful too.

File: tools/analyzers/recursivelock/testdata/complexlocks.go (hunk 3 above)
Comment (Heph789) [reply tools/analyzers/recursivelock/testdata/complexlocks.go:1]: Sure! I'll make it right now.

Comment (Heph789) [reply tools/analyzers/recursivelock/testdata/complexlocks.go:1]: #10078

**Review by nisdas — APPROVED** [review]
Great Work on this and finally having this ready ! This is going to be very useful for us.

**Pull request merged**